### PR TITLE
Added ranked buckets

### PR DESCRIPTION
**Pull request by AsafMah**
### Added
* Smarter way for retries with StorageAccounts

### CODE REVIEW

File: ingest/src/main/java/com/microsoft/azure/kusto/ingest/resources/RankedStorageAccount.java [hunk 1]
```diff
@@ -1,18 +1,90 @@
...
+        this.bucketCount = bucketCount;
+        this.bucketDurationInSec = bucketDurationInSec;
+        this.timeProvider = timeProvider;
+        lastActionTimestamp = timeProvider.currentTimeMillis();
     }
 
     public void addResult(boolean success) {
+        adjustForTimePassed();
+        Bucket lastBucket = buckets.peek();
+        assert lastBucket != null;
```
[the remark this answers is not in the document]
Comment (AsafMah) [reply ingest/src/main/java/com/microsoft/azure/kusto/ingest/resources/RankedStorageAccount.java:37]: adjustForTimePassed() always creates a bucked, the assert is for the IDE not to complain. I'll add a comment

File: ingest/src/main/java/com/microsoft/azure/kusto/ingest/resources/RankedStorageAccountSet.java [hunk 2]
```diff
@@ -1,17 +1,42 @@
...
+import com.microsoft.azure.kusto.ingest.utils.TimeProvider;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
 import java.util.*;
+import java.util.stream.Collectors;
 
 public class RankedStorageAccountSet {
+    private static final int DEFAULT_BUCKET_COUNT = 6;
+    private static final int DEFAULT_BUCKET_DURATION_IN_SEC = 10;
```
[the remark this answers is not in the document]
Comment (AsafMah) [reply ingest/src/main/java/com/microsoft/azure/kusto/ingest/resources/RankedStorageAccountSet.java:15]: 60 seconds per cycle can react faster to momentary changes. 
From experimenting it seems more helpful.

File: ingest/src/main/java/com/microsoft/azure/kusto/ingest/resources/RankedStorageAccountSet.java [hunk 3]
```diff
@@ -1,17 +1,42 @@
...
+    private static final int[] DEFAULT_TIERS = new int[] {90, 70, 30, 0};
```
[the remark this answers is not in the document]
Comment (AsafMah) [reply ingest/src/main/java/com/microsoft/azure/kusto/ingest/resources/RankedStorageAccountSet.java:16]: If an account is throttled and starts failing - you want to phase it out.

File: ingest/src/main/java/com/microsoft/azure/kusto/ingest/resources/RankedStorageAccount.java [hunk 4]
```diff
@@ -1,18 +1,93 @@
...
+    private Bucket adjustForTimePassed() {
+        if (buckets.isEmpty()) {
+            buckets.push(new Bucket());
+        }
+
+        long timePassed = timeProvider.currentTimeMillis() - lastActionTimestamp;
+        long bucketsToCreate = timePassed / (bucketDurationInSec * 1000L);
+        if (bucketsToCreate >= maxNumberOfBuckets) {
+            buckets.clear();
+            buckets.push(new Bucket());
```
Comment (alonadam) [inline ingest/src/main/java/com/microsoft/azure/kusto/ingest/resources/RankedStorageAccount.java:52]: Why is this needed? Not sure I understand.
Wouldn't it be polled anyway in line 58?

File: ingest/src/main/java/com/microsoft/azure/kusto/ingest/resources/RankedStorageAccount.java [hunk 5]
```diff
@@ -1,18 +1,93 @@
...
-    public RankedStorageAccount(String accountName) {
+    public RankedStorageAccount(String accountName, int maxNumberOfBuckets, int bucketDurationInSec, TimeProvider timeProvider) {
         this.accountName = accountName;
+        this.maxNumberOfBuckets = maxNumberOfBuckets;
+        this.bucketDurationInSec = bucketDurationInSec;
+        this.timeProvider = timeProvider;
+        lastActionTimestamp = timeProvider.currentTimeMillis();
     }
 
     public void addResult(boolean success) {
```
Comment (alonadam) [inline ingest/src/main/java/com/microsoft/azure/kusto/ingest/resources/RankedStorageAccount.java:34]: can be simplified and also makes more sense per getRank() calculations:
`

        Bucket lastBucket = adjustForTimePassed();

        lastBucket.total++;

        if (success) {

            lastBucket.successCount++;

        } 
`

File: ingest/src/main/java/com/microsoft/azure/kusto/ingest/resources/RankedStorageAccount.java [hunk 6]
```diff
@@ -1,18 +1,93 @@
...
+    private final int bucketDurationInSec;
+    private final TimeProvider timeProvider;
+
+    private long lastActionTimestamp;
 
-    public RankedStorageAccount(String accountName) {
+    public RankedStorageAccount(String accountName, int maxNumberOfBuckets, int bucketDurationInSec, TimeProvider timeProvider) {
         this.accountName = accountName;
+        this.maxNumberOfBuckets = maxNumberOfBuckets;
+        this.bucketDurationInSec = bucketDurationInSec;
```
[the remark this answers is not in the document]
Comment (AsafMah) [reply ingest/src/main/java/com/microsoft/azure/kusto/ingest/resources/RankedStorageAccount.java:29]: Good idea

File: ingest/src/main/java/com/microsoft/azure/kusto/ingest/resources/RankedStorageAccount.java (hunk 4 above)
Comment (AsafMah) [reply ingest/src/main/java/com/microsoft/azure/kusto/ingest/resources/RankedStorageAccount.java:52]: Good catch - the idea is to save computation time when we need to throw all the buckets away, there should have been an else in there, I'll fix.

File: ingest/src/main/java/com/microsoft/azure/kusto/ingest/resources/RankedStorageAccount.java (hunk 5 above)
Comment (AsafMah) [reply ingest/src/main/java/com/microsoft/azure/kusto/ingest/resources/RankedStorageAccount.java:34]: Great

**Review by alonadam — APPROVED** [review]
